docs/design/configs.md

```diff
@@ -0,0 +1,130 @@
+through the API server. ``Config`` resource will be mounted as a new volume 
+and will be managed by a side-car container that communicates with the API server.
+
+The ``Config`` resource architecture will be very similar to that of ``Secrets``.
+However, the ``Config`` resource will be dynamic, meaning that different components consuming
+the same ``Config`` will be notified about the changes to that particular resource.
+
+### Advantages
+
+* Reusable across different components.
+* Easy distribution through the API server.
+* By leveraging the power of ``/watch API`` along with the new resource, we can change component configurations dynamically.
+* Layer of abstraction that gets rid of the global state currently used for command-line flags and environment variables.
+
+### API Resource
+
+A new resource for ``Config`` will be added to the ``API``:
+
+```go
+	type Config struct {
+		TypeMeta   `json:",inline"`
+		ObjectMeta `json:"metadata,omitempty"`
+
+		// Flags contains command-line flags.
+		Flags pflag.FlagSet `json:"flags,omitempty"`
+
+		// Variables contains environment variables.
+		Env []EnvVar `json:"env,omitempty"`
+	}
+
+	type ConfigList struct {
+		TypeMeta `json:",inline"`
+		ListMeta `json:"metadata,omitempty"`
+
+		Items []Config `json:"items"`
+	}
+```
+``Registry`` interface will be added to accompany the ``Config`` resource.  
+``Config`` information will be stored in ``etcd`` by default.
+
+### Validation
+
+``Flags`` and ``Env`` are planned to be validated by type-checking at the moment for types provided
+beforehand for planned configuration parameters (``maxPodNumber`` => ``int``).
+
+### Volume Source
+
+A new ``ConfigSource`` type of volume source containing the ``Config`` object will be added to the ``VolumeSource`` struct in the API:
+
+```go
+	type ConfigVolumeSource struct {
+        ConfigName string `json:"configName"`
+	}
+```
+
+### Configuration
+
+Configured component (with ``Kubectl`` or possibly other means) will send it's configuration to the API server.
+
+```json
+{
+    "apiVersion": "v1beta3",
+    "kind": "Config",
+    "id": "",
+    "flags": {
+        "api_server": "132.32.13.14:8998",
+        "api_version": "v1beta3",
+    },
+    "variables": {
+        "DB_HOST": "1.1.1.1",
+        "DB_PORT": "8778",
+    }
+}
+```
+
+The owner of the configuration will be able mutate the configuration and it will be immutable from the
+perspective of other components (**Note:** A ``ConfigController`` might be added to undertake this responsibility).
+
+Consumers will have a container listening to the API server for configuration changes and reload them after a change.
+
+After the configuration is loaded the environment variables and flags will be visible inside the main container.
+The application will then be responsible from using the data.
+
+### Dynamic Configuration
+
+A side-car container will watch the API server and update the ``ConfigVolumeSource``.
```

Comment:
@tmrts I think if #5093 goes in, we can probably generalize a lot of it and avoid adding a lot of complexity to this proposal.  That said, I don't think there's any reason to have a volume plugin if you require a sidecar container to keep the volume updated.  I'm also concerned that a sidecar container will have data races that won't be an issue with the volume plugin approach because the volume plugin's `Setup` method will have run before the pod is started.